include/zephyr/sys/kref.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KREF_H_
+#define ZEPHYR_INCLUDE_SYS_KREF_H_
+
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup kref_apis Reference count APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when the count drops to 0. This can also be used on SMP systems.
+ */
+struct k_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize k_ref at compile-time. */
+#define K_REF_INIT()                                                                               \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize k_ref at runtime. */
+void k_ref_init(struct k_ref *ref);
```

Comment:
With these k_ref routines being as short as they are, I can't help but think that perhaps they should always be inlined.

Reply:
You're probably right. I wasn't sure about ref because it seemed rather "big" and the zephyr project seems to be all about optimizing stuff for a few bytes. But if we really end up simplifying ref by not protecting against refs from 0 then that argument goes away as well.

Reply:
I inlined them now